tests/ui/fn_param_redef_as_mutable.rs

```diff
@@ -0,0 +1,8 @@
+#![warn(clippy::fn_param_redef_as_mutable)]
+
+fn foobar(a: Vec<bool>) {
```

Comment:
We'll need more tests:

- more than one statement
- binding is not in the top level block
- arg is used after binding
- name of rebind is the same as name of arg
- name of rebind is the same as name of arg and arg is used afterwards.
- This test:

```rust
fn f(x: usize) {
    let y = x * 3;

    let x = 42;

    for i in 0..x {
        println!("{}", i);
    }

    let mut z = x; // Should not get linted
    z /= 2;

    println!("x={}, y={}, z={}", x, y, z);
}
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=f8986ee6bcfb8202f972af4645e71385)

Reply:
I noticed that you used arg instead of param, is there a preference to the wording here that I should use in the change?

Also just one thing that I think would be good for discussion; What's the intention behind not catching a rebind if they're not the same name of the arg?

Reply:
> What's the intention behind not catching a rebind if they're not the same name of the arg?

Oh, I should have specified, what of the above test cases should get linted and what not. I added it in my comment above

Reply:
> I noticed that you used arg instead of param

arg is shorter to write :smile: I don't know, what they are called in rust, honestly. I think, I heard both.

Reply:
Looking at the rustc doc, they are called param: https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.Param.html

Reply:
I've heard arguments are the _values_ that get passed to a function while parameters are the formal variables in the type signature. So in `fn f(a: usize)`, a is a parameter, but in `Vec::from("b")`, "b" is an argument.